verifiers/internal/gha/verifier.go

```diff
@@ -80,11 +72,72 @@ func (v *GHAVerifier) VerifyArtifact(ctx context.Context,
 	return r, builderID, err
 }
 
+// VerifyArtifact verifies provenance for an artifact.
+func (v *GHAVerifier) VerifyArtifact(ctx context.Context,
+	provenance []byte, artifactHash string,
+	provenanceOpts *options.ProvenanceOpts,
+	builderOpts *options.BuilderOpts,
+) ([]byte, string, error) {
+	rClient, err := rekor.NewClient(defaultRekorAddr)
+	if err != nil {
+		return nil, "", err
+	}
+
+	/* Verify signature on the intoto attestation. */
+	env, cert, err := VerifyProvenanceSignature(ctx, rClient, provenance, artifactHash)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return verifyEnvAndCert(env, cert, provenanceOpts, builderOpts)
+}
+
 // VerifyImage verifies provenance for an OCI image.
 func (v *GHAVerifier) VerifyImage(ctx context.Context,
-	provenance []byte, artifactHash string,
+	artifactReference string,
 	provenanceOpts *options.ProvenanceOpts,
 	builderOpts *options.BuilderOpts,
 ) ([]byte, string, error) {
-	return nil, "todo", serrors.ErrorNotSupported
+	/* Retrieve any valid signed attestations that chain up to Fulcio root CA. */
+	roots, err := fulcio.GetRoots()
+	if err != nil {
+		return nil, "", err
+	}
+	opts := &cosign.CheckOpts{
+		RootCerts: roots,
+	}
+
+	atts, _, err := container.RunCosignImageVerification(ctx,
+		artifactReference, opts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	/* Now verify properties of the attestations */
+	var verifyErr error
+	var builderID string
+	var verifiedProvenance []byte
+	for _, att := range atts {
+		pyld, err := att.Payload()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unexpected error getting payload from OCI registry %s", err)
+			continue
+		}
+		env, err := EnvelopeFromBytes(pyld)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unexpected error parsing envelope from OCI registry %s", err)
+			continue
+		}
+		cert, err := att.Cert()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unexpected error getting certificate from OCI registry %s", err)
+			continue
+		}
+		verifiedProvenance, builderID, verifyErr = verifyEnvAndCert(env, cert, provenanceOpts, builderOpts)
```

Comment:
do we want to verify the subject? Our builder uses `image:tag` as subject, so this should be safe?
/cc @ianlewis

Reply:
hmmm the subject digest is verified against the computed artifact digest. is this what you mean? the artiact digest is resolved and put into the `ProvenanceOpts`

Reply:
sorry. What I mean is that we don't verify the `subject.name`, but maybe we could since our generator puts the name of the container (including tags) in the name. That would provide a binding for container tags as well.

Reply:
Interesting, gotcha. In case the tags are different, we need to compare the resolved digest of the containers... which imo comes down to verifying digest anyway...

We could compare repository name?

I'm a little worried because I have to mock this out for tests anyway because local filepaths supply the pathname (not container name) as input.

It is easy to do though, we can pipe it through `ProvenanceOpts` and verify when we verify subject digest.

Reply:
There seems to be a few things to discuss, so maybe let's create an issue to discuss further?

Reply:
Done!
https://github.com/slsa-framework/slsa-verifier/issues/208